### PR TITLE
ENH: initial support a file-like backend with csv & hdf5 implementations     

### LINKED ISSUE #1165: HDF5 & CSV backends
I have an impl of these. placeholder issue.

### PR DESCRIPTION
**Pull request by jreback**
closes #1165

on top of #1149 (has needed compat)

### CODE REVIEW

File: ibis/file/client.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        return self.client.list_tables(path=path)
+
+
+@data_preload.register(HDFTable, pathlib.Path)
+def hdf_data_preload_uri_client(table, path, scope=None, **kwargs):
+    return pd.read_hdf(str(path), table.name, mode='r')
+
+
+@data_preload.register(CSVTable, pathlib.Path)
+def csv_data_preload_uri_client(table, path, scope=None, **kwargs):
```
Comment (jreback) [inline ibis/file/client.py]: here we could pretty easily push down a projection via ``usecols``

Comment (jreback) [reply ibis/file/client.py]: note we also assume ``index=False`` and a header in row 0 (as this matches how we write csvs). This could of course be more general, but makes sense to limit it for the easy case (and of course easy enough to simply allow passing in kwargs anyhow)

**Comment by jreback** [comment]
@cpcloud fixed up navigation, let's see if py2 issues are ok

File: ibis/file/csv/tests/test_basic.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def test_read_with_projection(csv2, data):
+
+    t = csv2.csv2.df
+    result = execute(t)
+    assert 'close' in result.columns
+    assert 'open' in result.columns
+
+    t = t[['time', 'ticker', 'close']]
```
Comment (jreback) [inline ibis/file/csv/tests/test_basic.py]: @cpcloud so this all works, but I'd like to intercept a projection call, *before* I pre-load, so that I can pass ``usecols`` in ``read_csv``.

I tried
```
@execute_node.register(ops.Selection, CSVTable)
def ....
```

but that's not right.

File: ibis/file/csv/tests/test_basic.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    # csv's don't preserve dtypes
+    expected = execute(t)
+    expected['time'] = expected['time'].astype(str)
+
+    tpath = tmpdir / 'new_csv'
+    tpath.mkdir()
+    path = tpath / 'foo.csv'
+
+    assert not path.exists()
+    c = ibis.csv.connect(tpath)
```
Comment (wesm) [inline ibis/file/csv/tests/test_basic.py]: What do you think about making this even simpler like `ibis.csv(path)` (or some other top level function name)?

Comment (cpcloud) [reply ibis/file/csv/tests/test_basic.py]: I kind of like this API because it's consistent with the rest of ibis therefore giving some indication of what to expect behavior-wise. Alternatively, what about making _all_ the backends look like this:

```python
ibis.impala()
ibis.postgres()
# etc...
```

Comment (wesm) [reply ibis/file/csv/tests/test_basic.py]: I'm OK with the `connect` API. In Spark it's `sc.read.parquet(...)` or `sc.read.csv`

**Comment by jreback** [comment]
@cpcloud ready for a look

File: ibis/__init__.py [hunk 4]
```diff
@@ -57,6 +57,17 @@
...
     pass
 
+try:
+    import ibis.file.csv.api as csv
+except ImportError:  # pip install  ibis-framework[csv]
+    pass
+
+
+try:
+    import ibis.file.hdf.api as hdf
```
Comment (cpcloud) [inline ibis/__init__.py]: Can we call this `hdf5` to avoid confusion with `hdfs`?

File: ibis/bigquery/tests/conftest.py [hunk 5]
```diff
@@ -11,7 +11,12 @@
 
 @pytest.fixture(scope='session')
 def client():
-    return ibis.bigquery.connect(PROJECT_ID, DATASET_ID)
+    ga = pytest.importorskip('google.auth')
```
Comment (cpcloud) [inline ibis/bigquery/tests/conftest.py]: You shouldn't need to use `importorskip` here (you can directly `import google.auth as ga`) since the `importorskip` for `google.cloud.bigquery` in `bigquery/tests/test_client.py` will run before this function is called.

File: ibis/file/client.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
+import ibis
+import ibis.expr.types as ir
+from ibis.pandas.execution import execute
+from ibis.file.utils import pathlib
+import ibis.file.execution  # noqa
+
+
+class Client(ibis.client.Client):
```
Comment (cpcloud) [inline ibis/file/client.py]: Let's call this `FileClient`

File: ibis/file/client.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        return execute(expr, params=params)
+
+    def list_tables(self, path=None):
+        raise NotImplementedError
+
+    def list_databases(self, path=None):
+        raise NotImplementedError
+
+
+class Database(ibis.client.Database):
```
Comment (cpcloud) [inline ibis/file/client.py]: `FileDatabase`?

File: ibis/file/client.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        super(Database, self).__init__(name, client)
+        self.path = path
+
+    def __str__(self):
+        return '{0.__class__.__name__}({0.name})'.format(self)
+
+    def __dir__(self):
+        dbs = self.list_databases(path=self.path)
+        tables = self.list_tables(path=self.path)
+        return sorted(list(set(dbs).union(set(tables))))
```
Comment (cpcloud) [inline ibis/file/client.py]: You don't need to sort this.

File: ibis/file/csv/api.py [hunk 9]
```diff
@@ -0,0 +1,5 @@
+from ibis.file.csv.client import CSVClient
+
+
+def connect(dictionary):
```
Comment (cpcloud) [inline ibis/file/csv/api.py]: Needs a small docstring describing what `dictionary` consists of since this is a user-facing API.

File: ibis/file/csv/tests/conftest.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    rng = pd.date_range('20170101', periods=10, freq='D')
+    tickers = ['GOOGL', 'FB', 'APPL', 'NFLX', 'AMZN']
+    return pd.DataFrame(
+        {'time': np.repeat(rng, len(tickers)),
+         'ticker': np.tile(tickers, len(rng))})
+
+
+@pytest.fixture
+def closes(df):
+    return df.assign(**{'close': np.random.randn(len(df))})
```
Comment (cpcloud) [inline ibis/file/csv/tests/conftest.py]: Can just directly assign, no? No need for `**`.

File: ibis/file/csv/tests/conftest.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+
+@pytest.fixture
+def opens(df):
+    return df.assign(**{'open': np.random.randn(len(df))})
```
Comment (cpcloud) [inline ibis/file/csv/tests/conftest.py]: Same comment as above, I don't think you need `**` here.

File: ibis/file/csv/tests/test_basic.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+from ibis.file.csv.client import CSVClient, CSVTable
+from ibis.file.client import Database, execute
+
+
+@pytest.fixture
+def transformed(csv):
+
+    # we need to cast to a timestamp type
+    # as we read in as strings
+    closes = csv.csv.close
```
Comment (cpcloud) [inline ibis/file/csv/tests/test_basic.py]: Can you call the second `csv` in `csv.csv` something else to indicate it's as directory like `csv_dir` or something like that?

File: ibis/file/csv/tests/test_basic.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+    # directory navigation
+    assert isinstance(csv, Database)
+    result = dir(csv)
+    assert result == ['csv']
+
+    prices = csv.csv
+    assert isinstance(prices, Database)
+    result = dir(prices)
+    assert result == ['close', 'open']
```
Comment (cpcloud) [inline ibis/file/csv/tests/test_basic.py]: Hm, not sure about this ... I like the idea of restricting, but this seems pretty restrictive. Maybe that's okay.

File: ibis/file/csv/tests/test_basic.py (hunk 2 above)
Comment (cpcloud) [reply ibis/file/csv/tests/test_basic.py]: One we implement `pre_execute` (or something like it) then you should be able to provide an implementation that allows you to traverse the expression and do this optimization.

File: ibis/file/hdf/api.py [hunk 14]
```diff
@@ -0,0 +1,5 @@
+from ibis.file.hdf.client import HDFClient
+
+
+def connect(dictionary):
```
Comment (cpcloud) [inline ibis/file/hdf/api.py]: Needs a small docstring describing what `dictionary` should contain.

File: ibis/file/hdf/client.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        self.dictionary[name] = path
+        return t
+
+    def list_tables(self, path=None):
+        # tables are individual tables within a file
+
+        if path is None:
+            path = self.root
+
+        if (path.is_file() and str(path).endswith(self.extension)):
```
Comment (cpcloud) [inline ibis/file/hdf/client.py]: Don't need parens here.

File: ibis/file/hdf/client.py [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+            with pd.HDFStore(str(path), mode='r') as store:
+                # strip leading /
+                return [k[1:] for k in store.keys()]
+
+        return []
+
+    def list_databases(self, path=None):
```
Comment (cpcloud) [inline ibis/file/hdf/client.py]: This looks very similar to the csv version. Is there a way to factor this out into `FileClient`?

File: ibis/file/hdf/tests/test_basic.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+pytest.importorskip('tables')
+
+from ibis.file.hdf.client import HDFClient, HDFTable  # noqa: E402
+from ibis.file.client import Database, execute  # noqa: E402
+
+
+@pytest.fixture
+def transformed(hdf):
+
+    closes = hdf.hdf.prices.close
```
Comment (cpcloud) [inline ibis/file/hdf/tests/test_basic.py]: Can you call the second `hdf` in `hdf.hdf` something else?

File: setup.py [hunk 18]
```diff
@@ -41,6 +41,8 @@
 pandas_requires = ['multipledispatch']
 clickhouse_requires = ['clickhouse-driver']
 bigquery_requires = ['google-cloud-bigquery']
+csv_requires = pandas_requires + ['tables>=3.0.0']
```
Comment (cpcloud) [inline setup.py]: This looks like it should be `hdf_requires` instead.

File: ibis/bigquery/tests/conftest.py (hunk 5 above)
Comment (jreback) [reply ibis/bigquery/tests/conftest.py]: note this is already pushed (in #1186). I can fix separately if you'd like.

File: ibis/file/client.py (hunk 6 above)
Comment (jreback) [reply ibis/file/client.py]: done

File: ibis/file/client.py (hunk 7 above)
Comment (jreback) [reply ibis/file/client.py]: done

File: ibis/file/client.py (hunk 8 above)
Comment (jreback) [reply ibis/file/client.py]: you don't but its nicer to be consistent on output

File: ibis/file/csv/api.py (hunk 9 above)
Comment (jreback) [reply ibis/file/csv/api.py]: done

File: ibis/file/csv/tests/conftest.py (hunk 10 above)
Comment (jreback) [reply ibis/file/csv/tests/conftest.py]: fixed; i originally did to avoid the highlite of open, but no big deal.

File: ibis/file/csv/tests/conftest.py (hunk 11 above)
Comment (jreback) [reply ibis/file/csv/tests/conftest.py]: fixed

File: ibis/file/csv/tests/test_basic.py (hunk 12 above)
Comment (jreback) [reply ibis/file/csv/tests/test_basic.py]: done

File: ibis/file/hdf/api.py (hunk 14 above)
Comment (jreback) [reply ibis/file/hdf/api.py]: done

File: ibis/file/hdf/tests/test_basic.py (hunk 17 above)
Comment (jreback) [reply ibis/file/hdf/tests/test_basic.py]: done

File: setup.py (hunk 18 above)
Comment (jreback) [reply setup.py]: done

File: ibis/file/hdf/client.py (hunk 15 above)
Comment (jreback) [reply ibis/file/hdf/client.py]: done

File: ibis/file/hdf/client.py (hunk 16 above)
Comment (jreback) [reply ibis/file/hdf/client.py]: its actually similar to the list_tables in csv. But databases in hdf5 are actually files, the tables are inside, so we have to traverse a bit differntly.

**Comment by jreback** [comment]
pushed

File: ibis/file/csv/tests/test_basic.py (hunk 2 above)
Comment (jreback) [reply ibis/file/csv/tests/test_basic.py]: yep

File: ibis/file/execution.py [hunk 19]
```diff
@@ -0,0 +1,11 @@
+from ibis.expr import operations as ops  # noqa
```
Comment (jreback) [inline ibis/file/execution.py]: so will remove this for now

File: ibis/file/client.py (hunk 8 above)
Comment (cpcloud) [reply ibis/file/client.py]: Right, I mean that it's already sorted for you so the consistency is there without having to provide it ourselves:

IPython:

```python
In [3]: class Foo:
   ...:     def __dir__(self):
   ...:         return ['b', 'a']
   ...: 

In [4]: foo = Foo()

In [5]: dir(foo)
Out[5]: ['a', 'b']
```

Vanilla Python:

```python
>>> class Foo:
...     def __dir__(self):
...         return ['b', 'a']
... 
>>> foo = Foo()
>>> dir(foo)
['a', 'b']
```

**Comment by jreback** [comment]
fails because of sorting on dir: https://circleci.com/gh/ibis-project/ibis/1508?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

File: ibis/file/client.py (hunk 8 above)
Comment (cpcloud) [reply ibis/file/client.py]: Ok, I'm not sure why this appears to always sort, but it appears to work on my machine and not necessarily others.

@jreback Keeping `sorted` is fine, but can you remove the call to `list` since `sorted` always returns a `list`?

File: ibis/file/csv/client.py [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class CSVTable(ops.DatabaseTable):
+    pass
+
+
+class CSVClient(FileClient):
+    extension = 'csv'
+
+    def write_to_file(self, path, t, index=False, **kwargs):
```
Comment (cpcloud) [inline ibis/file/csv/client.py]: Can we model this as an `insert` operation? So, instead of `write_to_table` the API is this:

```python
con = ibis.csv.connect('...')
con.create_table('path/to/file.csv')  # touch path/to/file.csv
expr = ...
con.insert('file.csv', expr)  # calls execute(expr).to_csv(...)
```

Comment (jreback) [reply ibis/file/csv/client.py]: yep

Comment (jreback) [reply ibis/file/csv/client.py]: though this is a bit redundant, I would simply do

```
con = ibis.csv.connect('/path/to')
expr = ...
con.insert('file.csv', expr)  # calls execute(expr).to_csv(...)
```
so the difference from sql is that a ``con`` here effectively has state, namely the directory level. We could not have this and instead do this as part of the ``.insert`` operation itself.

File: ibis/file/csv/client.py [hunk 21]
```diff
@@ -65,4 +66,26 @@ def list_databases(self, path=None):
 
 @data_preload.register(CSVTable, (pathlib.Path, str))
 def csv_data_preload_uri_client(table, path, scope=None, **kwargs):
-    return pd.read_csv(str(path), header=0)
+    if scope is not None:
+        usecols = scope.get('usecols')
+    else:
+        usecols = None
```
Comment (jreback) [inline ibis/file/csv/client.py]: @cpcloud @toryhaavik 

using ``pre_execute``!

File: ibis/file/csv/client.py [hunk 22]
```diff
@@ -65,4 +66,26 @@ def list_databases(self, path=None):
...
+def csv_pre_execute(op, **kwargs):
+
+    if not (isinstance(op.table.op(), CSVTable) and op.selections):
+        return {}
+
+    # get our source tables
+    pt = physical_tables(op.table.op())
+    valid = []
+    for t in pt:
+        valid.extend(t.schema.names)
```
Comment (jreback) [inline ibis/file/csv/client.py]: @cpcloud prob not very idiomatic here....

File: ibis/pandas/core.py [hunk 23]
```diff
@@ -168,6 +168,13 @@ def execute_without_scope(
     params = {k.op() if hasattr(k, 'op') else k: v for k, v in params.items()}
 
     new_scope = toolz.merge(scope, data_scope, params, factory=factory)
+
+    # pre_execute
```
Comment (jreback) [inline ibis/pandas/core.py]: this is ``execute_without_scope`` we need to do this here as well.

File: ibis/file/csv/client.py [hunk 24]
```diff
@@ -65,4 +66,26 @@ def list_databases(self, path=None):
...
-    return pd.read_csv(str(path), header=0)
+    if scope is not None:
+        usecols = scope.get('usecols')
+    else:
+        usecols = None
+    df = pd.read_csv(str(path), usecols=usecols, header=0)
+    return df
+
+
+@pre_execute.register(ops.Selection)
```
Comment (jreback) [inline ibis/file/csv/client.py]: I *really* want to register ``pre_execute`` on the ``ops.Selection(CSVTable)`` (rather than do this instance check). any way to do this?

File: ibis/pandas/core.py (hunk 23 above)
Comment (cpcloud) [reply ibis/pandas/core.py]: This already happens in `execute_with_scope` though. Why do we need to call it here?

File: ibis/file/csv/client.py (hunk 22 above)
Comment (cpcloud) [reply ibis/file/csv/client.py]: `scope` is really only for mapping expressions to data, not a generic way to passing keyword arguments. we should probably just use `**kwargs` for that kind of stuff.

File: ibis/file/csv/client.py (hunk 24 above)
Comment (cpcloud) [reply ibis/file/csv/client.py]: We can probably extend `pre_execute` to take a `client` argument, so that you don't have to do this instance checking. This would require a `_client` attribute on expressions or something like it.

File: ibis/pandas/core.py (hunk 23 above)
Comment (jreback) [reply ibis/pandas/core.py]: because this was getting called *first*, I am using ``execute(t)``, does that call ``execute_without_scope``?

I had a ``data_preload`` defined as well. These shouldn't be mutually exclusive. (and conceputally easier to put all of the data access in a single place as well).

File: ibis/file/csv/client.py (hunk 22 above)
Comment (jreback) [reply ibis/file/csv/client.py]: so should I actually do the physical read here then? (and just return ``{op:df}`` I think I tried that, let me try again.

File: ibis/pandas/core.py (hunk 23 above)
Comment (cpcloud) [reply ibis/pandas/core.py]: `execute_with_scope` calls `execute_without_scope`

maybe `data_preload` can go away if we add `client` to the dispatch. for now you should probably just define `pre_execute` since you want to do those optimizations.

File: ibis/file/csv/client.py (hunk 24 above)
Comment (jreback) [reply ibis/file/csv/client.py]: ok revised to just use ``pre_execute`` now.

but lots of hackiness because this is called pretty generally, IOW I only want to be called if I have a physical_table that is a CSVTable *in* the selection.

A tiny bit of hackiness with getting the path arg as well (because I can no longer dispatch on it), but that is ok.

File: ibis/file/csv/client.py [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+def csv_pre_execute_table(op, client, scope=None, **kwargs):
+    path = client.dictionary[op.name]
+    df = pd.read_csv(str(path), header=0)
+    return {op: df}
+
+
+@pre_execute.register(ops.Selection, CSVClient)
+def csv_pre_execute(op, client, scope=None, **kwargs):
+
+    pt = physical_tables(op.table.op())
```
Comment (jreback) [inline ibis/file/csv/client.py]: @cpcloud this works nicely now. not sure if its possible to have multiple physical tables here.

File: ibis/pandas/core.py [hunk 26]
```diff
@@ -85,8 +88,22 @@ def execute_with_scope(expr, scope, context=None, **kwargs):
...
+    # or cache it on the op?
+    try:
+        backends = find_backend(expr)
+        if not isinstance(backends, list):
+            backends = [backends]
+        for client in backends:
+            pre_loaded_scope = pre_execute(op,
+                                           client,
+                                           scope=scope,
+                                           context=context,
```
Comment (jreback) [inline ibis/pandas/core.py]: only needed a small change here

**Comment by jreback** [comment]
closing in lie of #1149 (contains everything)
